src/NLopt.jl

```diff
@@ -453,16 +453,23 @@ srand_time() = nlopt_srand_time()
 ############################################################################
 # Objective function:
 
+const _empty_vector = Cdouble[]
+
+@inline function _get_empty_vector()
+    @assert isempty(_empty_vector) "Builtin empty vector modified by user"
```

Comment:
Seems like it should have been a regular runtime check and exception rather than `@assert`, since we can't guarantee that this will always be `true`?

Reply:
I was initially thinking about it but felt like the most appropriate error to throw is `AssertionError`. I then realized that this can kind of be thought of as an internal consistency check if we claim that the user should not resize the array.

Reply:
That said, I'm fine with either way. If the grad input is supposed to be something that is undefined if modified by the user, then assert should be fine. If the grad is something that the user can mess with to get a well defined error then there should be a more meaningful error and also probably have a test for that. I slightly prefer the undefined behavior version since that's the only possible way in a multithreaded context.

Reply:
Yay or nay: https://github.com/jump-dev/NLopt.jl/pull/268?